### PR TITLE
Add service port mutation test

### LINKED ISSUE #110918: Missing Port mutation tests
### What happened?

We recently found that, when tweaking service objects manually, the following bug can exist (in alternative kube proxy implementations) , but it's not caught in the sig-network e2es... 

- make a service
- Confirm that the service works
- Edit the service by adding a new port to it
- Confirm both ports work
- Delete the added port
- Confirm these still work 

We need to
1. audit what port mutation tests we currently do
2. see what is supported
3. add tests for the missing deltas

Breaks.  However, we pass it seems, most or all of the sig-network e2es.  I think we have a missing e2e test.

Im pretty sure we support port mutation, right?  If not plz close this.

### What did you expect to happen?

e2e tests for sig-network would have coverage for service proxying robustness after port addition/deletion 

### How can we reproduce it (as minimally and precisely as possible)?

Reproducer is above, we  only noticed this lack of test coverage b/c of a broken service proxy :) 

### Anything else we need to know?

_No response_

### Kubernetes version

main

### Cloud provider

any

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by iXinqi**
#### What this PR does / why we need it:
Adding service port mutation test. More specifically:

- Create a service
- Confirm that the service works
- Edit the service by adding a new port to it
- Confirm both ports work
- Delete the added port
- Confirm these still work

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/110918

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by iXinqi** [comment]
Hi @jayunit100, may I have your review of this pr? I haven't tested it yet. Just want to make sure the test case meets our need. If it looks good to you, I'll go ahead and test it. Thanks.

**Comment by jayunit100** [comment]
hi xinqi - just compiled and tested it , appears to be working mostly (tested the negative case where my cluster was broke, and the test failed :)) 

lets make sure it passes other e2es...

/ok-to-test

File: test/e2e/network/service.go [hunk 1]
```diff
@@ -930,6 +930,109 @@ var _ = common.SIGDescribe("Services", func() {
 		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{})
 	})
 
+	ginkgo.It("should be able to add / delete ports ", func() {
```
Comment (jayunit100) [inline test/e2e/network/service.go:933]: `should be work for newly added and and deleted ports`

File: test/e2e/network/service.go [hunk 2]
```diff
@@ -930,6 +930,109 @@ var _ = common.SIGDescribe("Services", func() {
...
+				Name:          svc1port,
+				ContainerPort: int32(port1),
+			},
+		}
+		podname1 := "pod1"
+		createPodOrFail(f, ns, podname1, jig.Labels, containerPorts1, "netexec", "--http-port", strconv.Itoa(port1))
+		names[podname1] = true
+		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{podname1: {port1}}) // expect pod1 is added as ep
+
+		ginkgo.By("Checking if the Service forwards traffic to pods")
```
Comment (jayunit100) [inline test/e2e/network/service.go:978]: double checking,  any code duplication happening here that we should consolidate?

File: test/e2e/network/service.go [hunk 3]
```diff
@@ -930,6 +930,109 @@ var _ = common.SIGDescribe("Services", func() {
...
+			s.Spec.Ports = []v1.ServicePort{
+				{
+					Name:       "portname1",
+					Port:       80,
+					TargetPort: intstr.FromString(svc1port),
+				},
+			}
+		})
+		framework.ExpectNoError(err)
+
```
Comment (jayunit100) [inline test/e2e/network/service.go:1029]: should we also check here that port 80 fails?

**Comment by jayunit100** [comment]
Xinqi can you check into those failures ?

File: test/e2e/network/service.go (hunk 2 above)
Comment (iXinqi) [reply test/e2e/network/service.go]: Thanks for pointing this out. I changed the ginkgo.By msg so there is no duplicate code now. Does it look good to you now?

File: test/e2e/network/service.go (hunk 3 above)
Comment (iXinqi) [reply test/e2e/network/service.go:1029]: Yes, it's checked in the next two lines. If port 81 still exists, the actual port will be `portsByPodName{podname1: {port1},podname2: {port2}}` which is different from the expectation (`portsByPodName{podname1: {port1}}`), so the `validateEndpointsPortsOrFail()` will fail. Therefore, port 81 shouldn't even exist.

File: test/e2e/network/service.go [hunk 4]
```diff
@@ -930,6 +930,109 @@ var _ = common.SIGDescribe("Services", func() {
...
+	ginkgo.It("should be updated after adding or deleting ports ", func() {
+		// repacking functionality is intentionally not tested here - it's better to test it in an integration test.
+		serviceName := "edit-port-test"
+		ns := f.Namespace.Name
+		jig := e2eservice.NewTestJig(cs, ns, serviceName)
+
+		defer func() {
+			err := cs.CoreV1().Services(ns).Delete(context.TODO(), serviceName, metav1.DeleteOptions{})
+			framework.ExpectNoError(err, "failed to delete service: %s in namespace: %s", serviceName, ns)
+		}()
```
Comment (aojea) [inline test/e2e/network/service.go:942]: you don't need this defer, the e2e framework cleans up the namespace and the garbage collector delets the objects on it

File: test/e2e/network/service.go [hunk 5]
```diff
@@ -930,6 +930,109 @@ var _ = common.SIGDescribe("Services", func() {
...
+		framework.ExpectNoError(err)
+		validateEndpointsPortsOrFail(cs, ns, serviceName, portsByPodName{})
+
+		names := map[string]bool{}
+		defer func() {
+			for name := range names {
+				err := cs.CoreV1().Pods(ns).Delete(context.TODO(), name, metav1.DeleteOptions{})
+				framework.ExpectNoError(err, "failed to delete pod: %s in namespace: %s", name, ns)
+			}
+		}()
```
Comment (aojea) [inline test/e2e/network/service.go:964]: pods are deleted by the framework

**Comment by aojea** [comment]
lgtm. some nits to clean the code and stop carrying with the defer for cleanups on tests

File: test/e2e/network/service.go (hunk 4 above)
Comment (iXinqi) [reply test/e2e/network/service.go]: Got it, thanks!

**Comment by iXinqi** [comment]
Hi @aojea and @jayunit100, may I have your approval to merge this pr? Thanks!

**Comment by aojea** [comment]
great work, 
Thanks

/lgtm
/approve

**Pull request merged**